core/src/jac.jl

```diff
@@ -0,0 +1,495 @@
+            term_in = 1 / (area_in * R)
+            J[idx_in, idx_in] -= term_in
+        end
+
+        if has_index_out
+            area_out = basin.current_area[idx_out]
+            term_out = 1 / (area_out * R)
+            J[idx_out, idx_out] -= term_out
+        end
+
+        if has_index_in && has_index_out
+            J[idx_in, idx_out] += term_out
+            J[idx_out, idx_in] += term_in
+        end
+    end
+    return nothing
+end
+
+"""
+The contributions of ManningResistance nodes to the Jacobian.
+"""
+function formulate_jac!(
+    manning_resistance::ManningResistance,
+    J::SparseMatrixCSC{Float64, Int64},
+    u::ComponentVector{Float64},
+    p::Parameters,
+    t::Float64,
+)::Nothing
+    (; basin, connectivity) = p
+    (; node_id, active, length, manning_n, profile_width, profile_slope) =
+        manning_resistance
+    (; graph_flow) = connectivity
+
+    for (i, id) in enumerate(node_id)
+
+        # Inactive nodes do not contribute
+        if !active[i]
+            continue
+        end
+
+        #TODO: this was copied from formulate! for the manning_resistance,
+        # maybe put in separate function
+        basin_a_id = only(inneighbors(graph_flow, id))
+        basin_b_id = only(outneighbors(graph_flow, id))
+
+        h_a = get_level(p, basin_a_id)
+        h_b = get_level(p, basin_b_id)
+        bottom_a, bottom_b = basin_bottoms(basin, basin_a_id, basin_b_id, id)
+        slope = profile_slope[i]
+        width = profile_width[i]
+        n = manning_n[i]
+        L = length[i]
+
+        Δh = h_a - h_b
+        q_sign = sign(Δh)
+
+        # Average d, A, R
+        d_a = h_a - bottom_a
+        d_b = h_b - bottom_b
+        d = 0.5 * (d_a + d_b)
+
+        A_a = width * d + slope * d_a^2
+        A_b = width * d + slope * d_b^2
+        A = 0.5 * (A_a + A_b)
+
+        slope_unit_length = sqrt(slope^2 + 1.0)
+        P_a = width + 2.0 * d_a * slope_unit_length
+        P_b = width + 2.0 * d_b * slope_unit_length
+        R_h_a = A_a / P_a
+        R_h_b = A_b / P_b
+        R_h = 0.5 * (R_h_a + R_h_b)
+
+        k = 1000.0
+        kΔh = k * Δh
+        atankΔh = atan(k * Δh)
+        ΔhatankΔh = Δh * atankΔh
+        R_hpow = R_h^(2 / 3)
+        root = sqrt(2 / π * ΔhatankΔh)
+
+        q = q_sign * A / n * R_hpow * root / sqrt(L)
+
+        id_in = only(inneighbors(graph_flow, id))
+        id_out = only(outneighbors(graph_flow, id))
+
+        has_index_in, idx_in = id_index(basin.node_id, id_in)
+        has_index_out, idx_out = id_index(basin.node_id, id_out)
+
+        if has_index_in
+            basin_in_area = basin.current_area[idx_in]
+            ∂A_a = (width + 2 * slope * d_a) / basin_in_area
+            ∂A = 0.5 * ∂A_a
+            ∂P_a = 2 * slope_unit_length / basin_in_area
+            ∂R_h_a = (P_a * ∂A_a - A_a * ∂P_a) / P_a^2
+            ∂R_h_b = width / (2 * basin_in_area * P_b)
+            ∂R_h = 0.5 * (∂R_h_a + ∂R_h_b)
+            if Δh == 0
```

Comment:
In these cases I deliberatly didn't, because `sqrt_contribution` has a removable singularity, i.e. it doesn't exist at $\Delta h = 0$ because of division by zero but the limit $\Delta h \rightarrow 0$ does exist and is equal to the given value. I could either add a comment or indeed switch to $\approx$, for accuracy it probably doesn't matter much.

Reply:
Ok, thanks for clarifying, this is fine then.